source/base/hdf5.cc

```diff
@@ -0,0 +1,1513 @@
+      // the pointer will be wrong
+      return data.data();
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0];
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      void *>::type
+    get_container_pointer(Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0][0];
+    }
+
+    // This function returns the pointer to the raw data of a container
+    // The returned pointer is const, which means that it can be used only to
+    // read the data
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong
+      return data.data();
+    }
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, Vector<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &*data.begin();
+    }
+
+
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, FullMatrix<number>>::value,
+      const void *>::type
+    get_container_const_pointer(const Container<number> &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0][0];
+    }
+
+    // This function initializes a container of number type
+    template <template <class...> class Container, typename number>
+    typename std::enable_if<
+      std::is_same<Container<number>, std::vector<number>>::value,
+      Container<number>>::type
+    initialize_container(std::vector<hsize_t> dimensions)
```

Comment:
make it a reference

Reply:
Done

Reply:
nope, not done yet :-)

you can also make it `const`